### PR TITLE
Standard.likely and a new casting structure

### PR DESCRIPTION
**Pull request by JordanCheney**
Got stuck with some fast forwards so I made a new branch for this pull request. standard.likely is updated and a dummy cast with the new cast(Value *i, likely_hash hash) has been added

### CODE REVIEW

File: likely.cpp [hunk 1]
```diff
@@ -463,12 +463,21 @@ struct MatrixBuilder
...
+    Value *cast(Value *i, likely_hash hash) const { return (likely_type(h) == likely_type(hash)) ? i : b->CreateCast(CastInst::getCastOpcode(i, likely_is_signed(h), Type::getFloatTy(getGlobalContext()), likely_is_signed(h)), i, Type::getFloatTy(getGlobalContext())); }
     Value *add(Value *i, Value *j) const { return likely_is_floating(h) ? b->CreateFAdd(i, j, n) : b->CreateAdd(i, j, n); }
     Value *subtract(Value *i, Value *j) const { return likely_is_floating(h) ? b->CreateFSub(i, j, n) : b->CreateSub(i, j, n); }
     Value *multiply(Value *i, Value *j) const { return likely_is_floating(h) ? b->CreateFMul(i, j, n) : b->CreateMul(i, j, n); }
     Value *divide(Value *i, Value *j) const { return likely_is_floating(h) ? b->CreateFDiv(i, j, n) : (likely_is_signed(h) ? b->CreateSDiv(i,j, n) : b->CreateUDiv(i, j, n)); }
 
+    Value *log(Value *i) const { i = cast(i, likely_hash_f32); vector<Type *> args; args.push_back(Type::getFloatTy(getGlobalContext())); Function *f_log = Intrinsic::getDeclaration(TheModule, Intrinsic::log, args); return b->CreateCall(f_log, i, n); }
```
Comment (jklontz) [inline likely.cpp]: These functions all seem to share a lot of logic, can we factor out what they have in common?

File: likely.cpp [hunk 2]
```diff
@@ -689,8 +698,20 @@ class KernelBuilder
...
+                    else if (value == "log10") values.push_back(kernel.log10(operand));
+                    else if (value == "sin") values.push_back(kernel.sin(operand));
+                    else if (value == "cos") values.push_back(kernel.cos(operand));
+                    else if (value == "fabs") values.push_back(kernel.fabs(operand));
+                    else if (value == "sqrt") values.push_back(kernel.sqrt(operand));
+                    else if (value == "exp") values.push_back(kernel.exp(operand));
+                    else                     {likely_assert(false, "KernelBuilder::makeEquation unsupported operator: %s", value.c_str()); return NULL; }
                 } else {
-                    if (!likely_assert(values.size() >= 2, "KernelBuilder::make equation insufficient operands: %lu for operator: %s", values.size(), value.c_str())) return NULL;
+                    if (!likely_assert(values.size() >= 2, "KernelBuilder::make equation insufficient operands: %lu for operator: %s", values.size(), value.c_str())) return NULL;)
```
Comment (jklontz) [inline likely.cpp]: This tail parenthesis seems wrong?

**Comment by jklontz** [comment]
Excellent, this is what I hoped it would look like! Just a few minor comments, then we are ready to merge. Thanks!

**Comment by JordanCheney** [comment]
Fixed the trailing parenthesis (copy paste mishap) but casting is still not working properly

**Pull request merged**